[deploy] docs

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69251

This adds some actual documentation for deploy, which is probably useful
since we told everyone it was experimentally available so they will
probably be looking at what the heck it is.

It also wires up various compoenents of the OSS build to actually work
when used from an external project.

Differential Revision: [D32783312](https://our.internmc.facebook.com/intern/diff/D32783312)

suo committed:
[deploy] docs

This adds some actual documentation for deploy, which is probably useful
since we told everyone it was experimentally available so they will
probably be looking at what the heck it is.

It also wires up various compoenents of the OSS build to actually work
when used from an external project.

[ghstack-poisoned]

suo committed:
Update on "[deploy] docs"


This adds some actual documentation for deploy, which is probably useful
since we told everyone it was experimentally available so they will
probably be looking at what the heck it is.

It also wires up various compoenents of the OSS build to actually work
when used from an external project.

[ghstack-poisoned]

wconstab commented on docs/source/deploy.rst:
```diff
@@ -0,0 +1,179 @@
+
+    add_executable(example-app example-app.cpp)
+    target_link_libraries(example-app "${TORCH_LIBRARIES}")
+    set_property(TARGET example-app PROPERTY CXX_STANDARD 14)
```
is the CXX_STANDARD 14 related to deploy, or is that just boilerplate for any torch cmake project at this point?

suo replied:
just standard, I took this from our tests

wconstab commented on docs/source/deploy.rst:
```diff
@@ -0,0 +1,179 @@
+    std::cout << output.slice(/*dim=*/1, /*start=*/0, /*end=*/5) << '\n';
+
+Notably, the model's forward function is executing in Python, in an embedded
+CPython interpreter. The model object itself is responsible for load balancing
```
the model is not responsible for load balancing requests among interpreters, right?  The user needs to use their own request threads to make parallel requests into different interpreters.  I must be misunderstanding this part.

suo replied:
I meant that multiple threads can call the same model's forward method, and the ReplicatedObject internally dispatches to different interpreters. I'll edit to clarify.

wconstab reviewed (approved):
overall looks great, thanks for writing it up! Left a couple small questions.

suo committed:
Update on "[deploy] docs"


This adds some actual documentation for deploy, which is probably useful
since we told everyone it was experimentally available so they will
probably be looking at what the heck it is.

It also wires up various compoenents of the OSS build to actually work
when used from an external project.

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32783312).

suo committed:
Update on "[deploy] docs"


This adds some actual documentation for deploy, which is probably useful
since we told everyone it was experimentally available so they will
probably be looking at what the heck it is.

It also wires up various compoenents of the OSS build to actually work
when used from an external project.

Differential Revision: [D32783312](https://our.internmc.facebook.com/intern/diff/D32783312)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32783312).

suo committed:
Update on "[deploy] docs"


This adds some actual documentation for deploy, which is probably useful
since we told everyone it was experimentally available so they will
probably be looking at what the heck it is.

It also wires up various compoenents of the OSS build to actually work
when used from an external project.

Differential Revision: [D32783312](https://our.internmc.facebook.com/intern/diff/D32783312)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32783312).
